middleware/grpc/grpc.go

```diff
@@ -0,0 +1,120 @@
+// Package grpc implements an HTTP handler that responds to grpc checks.
+package grpc
+
+import (
+	"crypto/tls"
+	"log"
+	"net"
+
+	context "golang.org/x/net/context"
+	grpclib "google.golang.org/grpc"
+
+	"github.com/miekg/dns"
+	"github.com/miekg/coredns/core/dnsserver"
+	"github.com/miekg/coredns/middleware/grpc/pb"
+)
+
+type grpc struct {
+	addr   string
+	tls    *tls.Config
+	server *grpclib.Server
+	config *dnsserver.Config
+}
+
+func (g *grpc) Startup() error {
+	if g.addr == "" && g.tls == nil {
+		g.addr = defHttp
+	} else if g.addr == "" {
+		g.addr = defHttps
+	}
+
+	var ln net.Listener
+	var err error
+	if g.tls == nil {
+		ln, err = net.Listen("tcp", g.addr)
+	} else {
+		ln, err = tls.Listen("tcp", g.addr, g.tls)
+	}
+	if err != nil {
+		log.Printf("[ERROR] Failed to start grpc handler: %s", err)
+		return err
+	}
+
+	g.server = grpclib.NewServer()
+	pb.RegisterDnsServiceServer(g.server, g)
+	go func() {
+		g.server.Serve(ln)
+	}()
+	return nil
+}
+
+func (g *grpc) Query(ctx context.Context, in *pb.DnsPacket) (*pb.DnsPacket, error) {
+	msg := new(dns.Msg)
+	err := msg.Unpack(in.Msg)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO(jbelamaric): this is junk
```

Comment:
Not sure. I don't fully understand what it is used for. I could put the local address that gRPC is listening on and the remote address of the gRPC connection, I think. I think I can get that info from the context...but what's it for? can these be nil?